README.md
# asn1-ber-syntax

Type and functions for encoding and decoding ASN.1 documents. Look in
the test suite for examples of how to use this to encode and decode
SNMPv2 messages.


CHANGELOG.md
# Revision history for asn1-ber-syntax

## 0.2.0.0 -- 2022-04-11

* Support UtcTime and Boolean. Technically a breaking
  change since it adds data constructors.

## 0.1.0.0 -- 2022-??-??

* Initial release


test/Main.hs
{-# language DerivingStrategies #-}
{-# language GeneralizedNewtypeDeriving #-}
{-# language ScopedTypeVariables #-}
{-# language StandaloneDeriving #-}

import Asn.Ber (decode)
import Asn.Oid (Oid(Oid))
import Data.Bytes (Bytes)
import Test.Tasty.HUnit ((@=?))

import qualified Asn.Oid as Oid
import qualified Data.Bytes as Bytes
import qualified Test.Tasty as Tasty
import qualified Test.Tasty.HUnit as HUnit
import qualified Test.Tasty.QuickCheck as QC
import qualified Asn.Ber as Ber
import qualified GHC.Exts as Exts
import qualified Asn.Ber.Primitive.Decode as Decode
import qualified Asn.Ber.Primitive.Encode as Encode

main :: IO ()
main = Tasty.defaultMain $ Tasty.testGroup "tests"
  [ HUnit.testCase "signed-data" $ case decode exampleSignedData of
      Left e -> fail ("Decoding signed data failed with: " ++ show e)
      Right _ -> pure ()
  , HUnit.testCase "bad-integer-zero" $ case Decode.int64 badIntegerZero of
      Nothing -> pure ()
      Just _ -> fail ("Decoding bad integer zero succeeded unexpectedly.")
  , HUnit.testCase "good-integer-zero" $ case Decode.int64 goodIntegerZero of
      Nothing -> fail "Decoding good integer zero failed unexpectedly."
      Just 0 -> pure ()
      Just i -> fail ("Decoding good integer zero gave bad result: " ++ show i)
  , HUnit.testCase "oid-encode-001" $
      Exts.fromList [0x2a,0x86,0x48,0x86,0xf7,0x0d]
      @=?
      Encode.oid (Oid (Exts.fromList [1,2,840,113549]))
  , QC.testProperty "oid-encode-decode" $ \oid -> case Oid.size oid of
      0 -> True
      _ -> case Oid.fromShortTextDot (Oid.toShortText oid) of
        Nothing -> False
        Just r -> r == oid
  ]

badIntegerZero :: Bytes
badIntegerZero = Bytes.empty

goodIntegerZero :: Bytes
goodIntegerZero = Exts.fromList [0x00]

-- Taken from https://www.di-mgt.com.au/docs/examplesPKCS.txt
exampleSignedData :: Bytes
exampleSignedData = Exts.fromList
  [0x30,0x82,0x02,0x50,0x06,0x09,0x2a,0x86,0x48,0x86,0xf7,0x0d,0x01,0x07,0x02,0xa0
  ,0x82,0x02,0x41,0x30,0x82,0x02,0x3d,0x02,0x01,0x01,0x31,0x0e,0x30,0x0c,0x06,0x08
  ,0x2a,0x86,0x48,0x86,0xf7,0x0d,0x02,0x02,0x05,0x00,0x30,0x28,0x06,0x09,0x2a,0x86
  ,0x48,0x86,0xf7,0x0d,0x01,0x07,0x01,0xa0,0x1b,0x04,0x19,0x45,0x76,0x65,0x72,0x79
  ,0x6f,0x6e,0x65,0x20,0x67,0x65,0x74,0x73,0x20,0x46,0x72,0x69,0x64,0x61,0x79,0x20
  ,0x6f,0x66,0x66,0x2e,0xa0,0x82,0x01,0x5e,0x30,0x82,0x01,0x5a,0x30,0x82,0x01,0x04
  ,0x02,0x04,0x14,0x00,0x00,0x29,0x30,0x0d,0x06,0x09,0x2a,0x86,0x48,0x86,0xf7,0x0d
  ,0x01,0x01,0x02,0x05,0x00,0x30,0x2c,0x31,0x0b,0x30,0x09,0x06,0x03,0x55,0x04,0x06
  ,0x13,0x02,0x55,0x53,0x31,0x1d,0x30,0x1b,0x06,0x03,0x55,0x04,0x0a,0x13,0x14,0x45
  ,0x78,0x61,0x6d,0x70,0x6c,0x65,0x20,0x4f,0x72,0x67,0x61,0x6e,0x69,0x7a,0x61,0x74
  ,0x69,0x6f,0x6e,0x30,0x1e,0x17,0x0d,0x39,0x32,0x30,0x39,0x30,0x39,0x32,0x32,0x31
  ,0x38,0x30,0x36,0x5a,0x17,0x0d,0x39,0x34,0x30,0x39,0x30,0x39,0x32,0x32,0x31,0x38
  ,0x30,0x35,0x5a,0x30,0x42,0x31,0x0b,0x30,0x09,0x06,0x03,0x55,0x04,0x06,0x13,0x02
  ,0x55,0x53,0x31,0x1d,0x30,0x1b,0x06,0x03,0x55,0x04,0x0a,0x13,0x14,0x45,0x78,0x61
  ,0x6d,0x70,0x6c,0x65,0x20,0x4f,0x72,0x67,0x61,0x6e,0x69,0x7a,0x61,0x74,0x69,0x6f
  ,0x6e,0x31,0x14,0x30,0x12,0x06,0x03,0x55,0x04,0x03,0x13,0x0b,0x54,0x65,0x73,0x74
  ,0x20,0x55,0x73,0x65,0x72,0x20,0x31,0x30,0x5b,0x30,0x0d,0x06,0x09,0x2a,0x86,0x48
  ,0x86,0xf7,0x0d,0x01,0x01,0x01,0x05,0x00,0x03,0x4a,0x00,0x30,0x47,0x02,0x40,0x0a
  ,0x66,0x79,0x1d,0xc6,0x98,0x81,0x68,0xde,0x7a,0xb7,0x74,0x19,0xbb,0x7f,0xb0,0xc0
  ,0x01,0xc6,0x27,0x10,0x27,0x00,0x75,0x14,0x29,0x42,0xe1,0x9a,0x8d,0x8c,0x51,0xd0
  ,0x53,0xb3,0xe3,0x78,0x2a,0x1d,0xe5,0xdc,0x5a,0xf4,0xeb,0xe9,0x94,0x68,0x17,0x01
  ,0x14,0xa1,0xdf,0xe6,0x7c,0xdc,0x9a,0x9a,0xf5,0x5d,0x65,0x56,0x20,0xbb,0xab,0x02
  ,0x03,0x01,0x00,0x01,0x30,0x0d,0x06,0x09,0x2a,0x86,0x48,0x86,0xf7,0x0d,0x01,0x01
  ,0x02,0x05,0x00,0x03,0x41,0x00,0x45,0x1a,0xa1,0xe1,0xaa,0x77,0x20,0x4a,0x5f,0xcd
  ,0xf5,0x76,0x06,0x9d,0x02,0xf7,0x32,0xc2,0x6f,0x36,0x7b,0x0d,0x57,0x8a,0x6e,0x64
  ,0xf3,0x9a,0x91,0x1f,0x47,0x95,0xdf,0x09,0x94,0x34,0x05,0x11,0xa0,0xd1,0xdf,0x4a
  ,0x20,0xb2,0x6a,0x77,0x4c,0xca,0xef,0x75,0xfc,0x69,0x2e,0x54,0xc2,0xa1,0x93,0x7c
  ,0x07,0x11,0x26,0x9d,0x9b,0x16,0x31,0x81,0x9b,0x30,0x81,0x98,0x02,0x01,0x01,0x30
  ,0x34,0x30,0x2c,0x31,0x0b,0x30,0x09,0x06,0x03,0x55,0x04,0x06,0x13,0x02,0x55,0x53
  ,0x31,0x1d,0x30,0x1b,0x06,0x03,0x55,0x04,0x0a,0x13,0x14,0x45,0x78,0x61,0x6d,0x70
  ,0x6c,0x65,0x20,0x4f,0x72,0x67,0x61,0x6e,0x69,0x7a,0x61,0x74,0x69,0x6f,0x6e,0x02
  ,0x04,0x14,0x00,0x00,0x29,0x30,0x0c,0x06,0x08,0x2a,0x86,0x48,0x86,0xf7,0x0d,0x02
  ,0x02,0x05,0x00,0x30,0x0d,0x06,0x09,0x2a,0x86,0x48,0x86,0xf7,0x0d,0x01,0x01,0x01
  ,0x05,0x00,0x04,0x40,0x05,0xfa,0x6a,0x81,0x2f,0xc7,0xdf,0x8b,0xf4,0xf2,0x54,0x25
  ,0x09,0xe0,0x3e,0x84,0x6e,0x11,0xb9,0xc6,0x20,0xbe,0x20,0x09,0xef,0xb4,0x40,0xef
  ,0xbc,0xc6,0x69,0x21,0x69,0x94,0xac,0x04,0xf3,0x41,0xb5,0x7d,0x05,0x20,0x2d,0x42
  ,0x8f,0xb2,0xa2,0x7b,0x5c,0x77,0xdf,0xd9,0xb1,0x5b,0xfc,0x3d,0x55,0x93,0x53,0x50
  ,0x34,0x10,0xc1,0xe1]

instance QC.Arbitrary Oid where
  arbitrary = Oid.Oid . Exts.fromList <$> QC.arbitrary


test/Golden.hs
import Asn.Ber (decode)
import Message (resolveMessage)
import Data.ByteString (ByteString)
import Data.ByteString.Base16 (decodeBase16)
import Data.Bytes (Bytes)
import Data.Char (isSpace)
import System.FilePath (takeBaseName, replaceExtension)
import System.IO (withFile,IOMode(WriteMode))
import Test.Tasty (defaultMain, TestTree, testGroup)
import Test.Tasty.Golden (goldenVsFile, findByExtension)
import Text.Pretty.Simple (pHPrint)

import qualified Data.ByteString as ByteString
import qualified Data.ByteString.Char8 as BC8
import qualified GHC.Exts as Exts
import qualified Message.Category as Cat

main :: IO ()
main = defaultMain =<< goldenTests

goldenTests :: IO TestTree
goldenTests = do
  a <- goldenTestsSchemaless
  b <- goldenResolveTests
  pure (testGroup "tests" [a,b])

goldenTestsSchemaless :: IO TestTree
goldenTestsSchemaless = do
  files <- findByExtension [".input"] "golden/schemaless"
  return $ testGroup "ber" $ flip map files $ \inputName -> do
    let actualName = replaceExtension inputName ".actual"
    let expectedName = replaceExtension inputName ".expected"
    goldenVsFile (takeBaseName inputName) expectedName actualName $ do
      b16 <- ByteString.readFile inputName
      case decodeBase16 (BC8.filter (not . isSpace) b16) of
        Left _ -> fail "file contained non-hexadecimal characters"
        Right b -> case decode (bytestringToBytes b) of
          Left err -> fail ("could not decode input: " ++ err)
          Right s -> withFile actualName WriteMode (\h -> pHPrint h s)

goldenResolveTests :: IO TestTree
goldenResolveTests = do
  files <- findByExtension [".input-resolver"] "golden/snmp-message"
  return $ testGroup "resolve" $ flip map files $ \inputName -> do
    let actualName = replaceExtension inputName ".actual"
    let expectedName = replaceExtension inputName ".expected"
    goldenVsFile (takeBaseName inputName) expectedName actualName $ do
      b16 <- ByteString.readFile inputName
      case decodeBase16 (BC8.filter (not . isSpace) b16) of
        Left _ -> fail "file contained non-hexadecimal characters"
        Right b -> case decode (bytestringToBytes b) of
          Left err -> fail ("could not decode input: " ++ err)
          Right s -> case resolveMessage s of
            Left err -> fail ("could not resolve input: " ++ show err)
            Right _ -> case Cat.resolveMessage s of
              Left err -> fail ("could not resolve input (Category-style): " ++ show err)
              Right v -> withFile actualName WriteMode (\h -> pHPrint h v)


bytestringToBytes :: ByteString -> Bytes
bytestringToBytes = Exts.fromList . ByteString.unpack


test/Message.hs
{-# LANGUAGE ApplicativeDo #-}
{-# LANGUAGE NamedFieldPuns #-}

module Message
  ( resolveMessage
  , Message(..)
  ) where

import Prelude hiding (sequence)

import Asn.Oid (Oid)
import Asn.Resolve

import Data.Bytes (Bytes)
import Data.Int (Int64)
import Data.Primitive (SmallArray)

data Message = Message
  { version :: Int64
  , community :: {-# UNPACK #-} !Bytes
  , pdu :: Pdu
  }
  deriving(Show)
data Pdu
  = GetRequest APdu
  | GetNextRequest APdu
  | Response APdu
  | SetRequest APdu
  | InformRequest APdu
  | SnmpV2Trap APdu
  | Report APdu
  deriving(Show)
data APdu = Pdu
  { requestId :: Int64
  , errorStatus :: Int64
  , errorIndex :: Int64
  , varBinds :: SmallArray VarBind
  }
  deriving(Show)
data VarBind = VarBind
  { name :: !Oid
  , result :: !VarBindResult
  }
  deriving(Show)
data VarBindResult
  = Value ObjectSyntax
  | Unspecified
  | NoSuchObject
  | NoSuchInstance
  | EndOfMibView
  deriving(Show)
data ObjectSyntax
  = IntegerValue Int64
  | StringValue Bytes
  | ObjectIdValue Oid
  | IpAddressValue Bytes
  | CounterValue Int64
  | TimeticksValue Int64
  | ArbitraryValue Bytes
  | BigCounterValue Int64
  | UnsignedIntegerValue Int64
  deriving(Show)

resolveMessage :: Value -> Either Path Message
resolveMessage = run . message

message :: Value -> Parser Message
message = sequence $ do
  version <- index 0 int64
  community <- index 1 octetString
  pdu <- index 2 resolvePdu
  pure Message{version,community,pdu}
  where
  resolvePdu = chooseTag
    [ (ContextSpecific, 0, fmap GetRequest . aPdu)
    , (ContextSpecific, 1, fmap GetNextRequest . aPdu)
    , (ContextSpecific, 3, fmap Response . aPdu)
    , (ContextSpecific, 4, fmap SetRequest . aPdu)
    , (ContextSpecific, 5, fmap InformRequest . aPdu)
    , (ContextSpecific, 6, fmap SnmpV2Trap . aPdu)
    , (ContextSpecific, 7, fmap Report . aPdu)
    ]
  aPdu :: Value -> Parser APdu
  aPdu = sequence $ do
    requestId <- index 0 int64
    errorStatus <- index 1 int64
    errorIndex <- index 2 int64
    varBinds <- index 3 $ sequenceOf $ sequence $ do
      name <- index 0 oid
      result <- index 1 varBind
      pure VarBind{name,result}
    pure Pdu{requestId,errorStatus,errorIndex,varBinds}
  varBind :: Value -> Parser VarBindResult
  varBind = chooseTag
    [(Application, 1, fmap (Value . CounterValue) . int64)]
    -- TODO


src/Asn/Ber.hs
{-# language BangPatterns #-}
{-# language BinaryLiterals #-}
{-# language DerivingStrategies #-}
{-# language LambdaCase #-}
{-# language MagicHash #-}
{-# language MultiWayIf #-}
{-# language NamedFieldPuns #-}
{-# language NumericUnderscores #-}
{-# language PatternSynonyms #-}
{-# language RankNTypes #-}
{-# language TypeApplications #-}
{-# language UnboxedTuples #-}

module Asn.Ber
  ( Value(..)
  , Contents(..)
  , Class(..)
  , decode
    -- * Constructed Patterns
  , pattern Set
  , pattern Sequence
  ) where

import Asn.Oid (Oid(..))
import Control.Monad (when)
import Data.Bits ((.&.),(.|.),testBit,unsafeShiftR,unsafeShiftL,complement)
import Data.Bytes (Bytes)
import Data.Bytes.Parser (Parser)
import Data.ByteString.Short.Internal (ShortByteString(SBS))
import Data.Int (Int64)
import Data.Primitive (SmallArray)
import Data.Word (Word8,Word32)
import GHC.Exts (Int(I#))
import GHC.ST (ST(ST))

import qualified Chronos
import qualified Data.Bytes as Bytes
import qualified Data.Bytes.Parser as P
import qualified Data.Bytes.Parser.Latin as Latin
import qualified Data.Bytes.Parser.Base128 as Base128
import qualified Data.Primitive as PM
import qualified Data.Text.Short as TS
import qualified Data.Text.Short.Unsafe as TS
import qualified GHC.Exts as Exts

data Value = Value
  { tagClass :: !Class
  , tagNumber :: !Word32
  , contents :: !Contents
  }
  deriving stock (Show)
  deriving stock (Eq)

data Contents
  -- x = Boolean !Bool
  -- x   -- ^ Tag number: @0x01@
  -- x | Integer !Int64
  -- x   -- ^ Tag number: @0x02@
  -- x | OctetString {-# UNPACK #-} !Bytes
  -- x   -- ^ Tag number: @0x04@
  -- x | BitString !Word8 {-# UNPACK #-} !Bytes
  -- x   -- ^ Tag number: @0x03@. Has padding bit count and raw bytes.
  -- x | Null
  -- x   -- ^ Tag number: @0x05@
  -- x | ObjectIdentifier !Oid
  -- x   -- ^ Tag number: @0x06@
  -- x | Utf8String {-# UNPACK #-} !TS.ShortText
  -- x   -- ^ Tag number: @0x0C@
  -- x | PrintableString {-# UNPACK #-} !TS.ShortText
  -- x   -- ^ Tag number: @0x13@
  -- x | UtcTime !Int64
  -- x   -- ^ Tag number: @0x17@. Number of seconds since the epoch.
  -- x   -- The following guidance is inspired by RFC 5280:
  -- x   --
  -- x   -- * A two-digit year greater than or equal to 50 is interpreted
  -- x   --   as 19XX, and a two-digit year less than 50 is intepreted
  -- x   --   as 20XX.
  -- x   -- * Everything is converted to Zulu time zone. Unlike RFC 5280,
  -- x   --   we do not require Zulu, but we convert everything to it.
  -- x   -- * When seconds are absent, we treat the timestamp as one where
  -- x   --   the seconds are zero. That is, we understand 2303252359Z as
  -- x   --   2023-03-25T23:59:00Z.
  = Constructed {-# UNPACK #-} !(SmallArray Value)
    -- ^ Constructed value contents in concatenation order.
    -- The class and tag are held in `Value`.
  | Primitive {-# UNPACK #-} !Bytes
    -- ^ Values that require information about interpreting application,
    -- context-specific, or private tag.
  deriving stock (Show)
  deriving stock (Eq)

pattern Sequence :: Word32
pattern Sequence = 0x10

pattern Set :: Word32
pattern Set = 0x11

data Class
  = Universal
  | Application
  | ContextSpecific
  | Private
  deriving stock (Eq,Show)

decode :: Bytes -> Either String Value
decode = P.parseBytesEither parser

decodePayload :: (forall s. Word -> Parser String s a) -> Bytes -> Either String a
decodePayload k bs =
  let len = fromIntegral @Int @Word (Bytes.length bs)
   in P.parseBytesEither (k len) bs

takeLength :: Parser String s Word
takeLength = do
  w <- P.any "tried to take the length"
  case testBit w 7 of
    False -> pure (fromIntegral w)
    True -> do
      let go !n !acc = case n of
            0 -> pure acc
            _ -> if acc < 16_000_000
              then do
                x <- P.any "while taking length, ran out of bytes"
                let acc' = fromIntegral @Word8 @Word x + (acc * 256)
                go (n - 1) acc'
              else P.fail "that is a giant length, bailing out"
      go (fromIntegral @Word8 @Word w .&. 0b01111111) 0

primitive :: Parser String s Contents
primitive = do
  n <- takeLength
  bs <- P.take "while decoding unresolved contents, not enough bytes" (fromIntegral n)
  pure (Primitive bs)

constructed :: Parser String s Contents
constructed = do
  n <- takeLength
  P.delimit "constructed not enough bytes" "constructed leftovers" (fromIntegral n) $ do
    let initialSize = 8
    buf0 <- P.effect (PM.newSmallArray initialSize errorThunk)
    let go !ix !sz !buf = P.isEndOfInput >>= \case
          True -> do
            res <- P.effect $ do
              buf' <- resizeSmallMutableArray buf ix
              PM.unsafeFreezeSmallArray buf'
            pure (Constructed res)
          False -> if ix < sz
            then do
              v <- parser
              P.effect (PM.writeSmallArray buf ix v)
              go (ix + 1) sz buf
            else do
              let newSz = sz * 2
              newBuf <- P.effect $ do
                newBuf <- PM.newSmallArray newSz errorThunk
                PM.copySmallMutableArray newBuf 0 buf 0 sz
                pure newBuf
              go ix newSz newBuf
    go 0 initialSize buf0

resizeSmallMutableArray :: PM.SmallMutableArray s a -> Int -> ST s (PM.SmallMutableArray s a)
resizeSmallMutableArray (PM.SmallMutableArray x) (I# i) =
  ST (\s -> (# Exts.shrinkSmallMutableArray# x i s, PM.SmallMutableArray x #))

errorThunk :: a
{-# noinline errorThunk #-}
errorThunk = errorWithoutStackTrace "Asn.Ber: implementation mistake"


classFromUpperBits :: Word8 -> Class
classFromUpperBits w = case unsafeShiftR w 6 of
  0 -> Universal
  1 -> Application
  2 -> ContextSpecific
  _ -> Private

parser :: Parser String s Value
parser = do
  b <- P.any "expected tag byte"
  let tagClass = classFromUpperBits b
      isConstructed = testBit b 5
  tagNumber <- case b .&. 0b00011111 of
        31 -> Base128.word32 "bad big tag"
        num -> pure $ fromIntegral @Word8 @Word32 num
  contents <- if isConstructed then constructed else primitive
  pure Value{tagClass, tagNumber, contents}

ba2stUnsafe :: PM.ByteArray -> TS.ShortText
ba2stUnsafe (PM.ByteArray x) = TS.fromShortByteStringUnsafe (SBS x)

ba2sbs :: PM.ByteArray -> ShortByteString
ba2sbs (PM.ByteArray x) = SBS x


src/Asn/Ber/Primitive/Decode.hs
{-# language BangPatterns #-}
{-# language DataKinds #-}
{-# language LambdaCase #-}
{-# language NumericUnderscores #-}
{-# language TypeApplications #-}

module Asn.Ber.Primitive.Decode
  ( int64
  , nonnegativeIntegerBytes20
  , utcTime
  , oid
  , null
  , boolean
  , bitString
  , bitStringBytes
  , octetString
  , utf8String
  , printableString
  ) where

import Prelude hiding (null)

import Asn.Oid (Oid(Oid))
import Data.Bytes (Bytes)
import Data.Int (Int64)
import Data.Bits (unsafeShiftL,complement,testBit,(.|.))
import Data.Word (Word8,Word32)
import Data.Bytes.Types (ByteArrayN(ByteArrayN))
import Data.ByteString.Short.Internal (ShortByteString(SBS))
import Control.Monad.ST.Run (runByteArrayST)
import Data.Bytes.Parser (Parser)
import Data.Text.Short (ShortText)

import qualified Data.Bytes.Parser.Base128 as Base128
import qualified Data.Bytes as Bytes
import qualified Data.Bytes.Parser.Latin as Latin
import qualified Data.Bytes.Parser as P
import qualified Data.Primitive as PM
import qualified Data.Text.Short as TS
import qualified Data.Text.Short.Unsafe as TS
import qualified Chronos

-- | Universal tag number: @0x01@. Expects boolean encoding.
boolean :: Bytes -> Maybe Bool 
boolean !b = case Bytes.length b of
  1 -> Just $! Bytes.unsafeIndex b 0 /= 0
  _ -> Nothing

-- | Universal tag number: @0x04@. Expects octet string encoding.
octetString :: Bytes -> Maybe Bytes
octetString = Just

-- | Universal tag number: @0x0C@. Expects utf8 string encoding.
utf8String :: Bytes -> Maybe ShortText
utf8String = TS.fromShortByteString . Bytes.toShortByteString

-- | Universal tag number: @0x13@. Expects printable string encoding.
printableString :: Bytes -> Maybe ShortText
printableString bs =
  if Bytes.all isPrintable bs
    then pure $! ba2stUnsafe $! Bytes.toByteArrayClone bs
    else Nothing

ba2stUnsafe :: PM.ByteArray -> TS.ShortText
ba2stUnsafe (PM.ByteArray x) = TS.fromShortByteStringUnsafe (SBS x)

-- | Universal tag number: @0x02@. Expects integer encoding.
int64 :: Bytes -> Maybe Int64
int64 !content = case Bytes.length content of
  -- Empty input is not an acceptable way to encode the number zero.
  0 -> Nothing
  _ ->
    let isNegative = testBit (Bytes.unsafeIndex content 0) 7
        loopBody acc b = (acc `unsafeShiftL` 8) .|. fromIntegral @Word8 @Int64 b
        seed = if isNegative then complement 0 else (0 :: Int64)
     in Just $! Bytes.foldl' loopBody seed content

-- | Universal tag number: @0x02@. Expects integer encoding. If the
-- integer is negative, this rejects it.
-- If it is a number less than @256^20@, accept it. Integers with a
-- small number of bytes are padded with zeros on the left.
nonnegativeIntegerBytes20 :: Bytes -> Maybe (ByteArrayN 20)
nonnegativeIntegerBytes20 !b = case Bytes.length b of
  -- Empty input is not an acceptable way to encode the number zero.
  0 -> Nothing
  len
    | w0 <- Bytes.unsafeIndex b 0 -> case len of
        1 -> if testBit w0 7
          then Nothing -- reject negative number
          else Just $! ByteArrayN $ runByteArrayST $ do
            dst <- PM.newByteArray 20
            PM.setByteArray dst 0 20 (0 :: Word8)
            PM.writeByteArray dst 19 w0
            PM.unsafeFreezeByteArray dst
        _ | w1 <- Bytes.unsafeIndex b 1 -> if testBit w0 7
              then Nothing -- reject negative number
              else case w0 of
                0 -> if testBit w1 7
                  then if len <= 21
                    then Just $! ByteArrayN $ runByteArrayST $ do
                      dst <- PM.newByteArray 20
                      PM.setByteArray dst 0 20 (0 :: Word8)
                      Bytes.unsafeCopy dst (21 - len) (Bytes.unsafeDrop 1 b)
                      PM.unsafeFreezeByteArray dst
                    else Nothing -- more than 20 bytes of data
                  else Nothing -- reject nine leading zero bits
                _ -> if len <= 20
                  then Just $! ByteArrayN $ runByteArrayST $ do
                    dst <- PM.newByteArray 20
                    PM.setByteArray dst 0 20 (0 :: Word8)
                    Bytes.unsafeCopy dst (20 - len) b
                    PM.unsafeFreezeByteArray dst
                  else Nothing -- more than 20 bytes of data

-- | Universal tag number: @0x17@. Expects @utctime@ encoding.
-- If the seconds part is missing, treating it as zero seconds past the
-- minute. A timestamp with a non-Zulu time zone is normalized by
-- converting to Zulu time. Returns the number of seconds since the epoch.
-- The following guidance is inspired by RFC 5280:
--
-- * A two-digit year greater than or equal to 50 is interpreted
--   as 19XX, and a two-digit year less than 50 is intepreted
--   as 20XX.
-- * Everything is converted to Zulu time zone. Unlike RFC 5280,
--   we do not require Zulu, but we convert everything to it.
-- * When seconds are absent, we treat the timestamp as one where
--   the seconds are zero. That is, we understand 2303252359Z as
--   2023-03-25T23:59:00Z.
utcTime :: Bytes -> Maybe Int64
utcTime = P.parseBytesMaybe (utcTimeParser <* P.endOfInput "utctime expected eof")

utcTimeParser :: Parser String s Int64
utcTimeParser = do
  !year0 <- twoDigits "utctime year digit problem"
  let !year = if year0 >= 50 then 1900 + year0 else 2000 + year0
  !month <- twoDigits "utctime month digit problem"
  !day <- twoDigits "utctime day digit problem"
  !hour <- twoDigits "utctime hour digit problem"
  !minute <- twoDigits "utctime minute digit problem"
  -- Offset must be provided in seconds.
  let finishWithoutSeconds !offset = case Chronos.timeFromYmdhms year month day hour minute 0 of
        Chronos.Time ns -> pure $! offset + div ns 1_000_000_000
  let finishWithSeconds !offset !seconds = case Chronos.timeFromYmdhms year month day hour minute seconds of
        Chronos.Time ns -> pure $! offset + div ns 1_000_000_000
  Latin.peek >>= \case
    Nothing -> finishWithoutSeconds 0
    Just c -> case c of
      'Z' -> do
        _ <- P.any "utctime impossible"
        finishWithoutSeconds 0
      '+' -> do
        _ <- P.any "utctime impossible"
        !offsetHour <- twoDigits "utctime offset hour digit problem"
        !offsetMinute <- twoDigits "utctime offset minute digit problem"
        let !offset = fromIntegral @Int @Int64 (negate (60 * (60 * offsetHour + offsetMinute)))
        finishWithoutSeconds offset
      '-' -> do
        _ <- P.any "utctime impossible"
        !offsetHour <- twoDigits "utctime offset hour digit problem"
        !offsetMinute <- twoDigits "utctime offset minute digit problem"
        let !offset = fromIntegral @Int @Int64 (60 * (60 * offsetHour + offsetMinute))
        finishWithoutSeconds offset
      _ | c >= '0', c <= '9' -> do
            seconds <- twoDigits "utctime seconds digit problem"
            Latin.peek >>= \case
              Nothing -> finishWithSeconds 0 seconds
              Just d -> case d of
                'Z' -> do
                  _ <- P.any "utctime impossible"
                  finishWithSeconds 0 seconds
                '+' -> do
                  _ <- P.any "utctime impossible"
                  !offsetHour <- twoDigits "utctime offset hour digit problem"
                  !offsetMinute <- twoDigits "utctime offset minute digit problem"
                  let !offset = fromIntegral @Int @Int64 (negate (60 * (60 * offsetHour + offsetMinute)))
                  finishWithSeconds offset seconds
                '-' -> do
                  _ <- P.any "utctime impossible"
                  !offsetHour <- twoDigits "utctime offset hour digit problem"
                  !offsetMinute <- twoDigits "utctime offset minute digit problem"
                  let !offset = fromIntegral @Int @Int64 (60 * (60 * offsetHour + offsetMinute))
                  finishWithSeconds offset seconds
                _ -> P.fail "utctime unexpected byte after seconds"
      _ -> P.fail "utctime unexpected byte after minute"

twoDigits :: e -> Parser e s Int
{-# inline twoDigits #-}
twoDigits e = do
  w0 <- P.any e
  w0' <- if w0 >= 0x30 && w0 <= 0x39
    then pure (fromIntegral @Word8 @Int w0 - 0x30)
    else P.fail e
  w1 <- P.any e
  w1' <- if w1 >= 0x30 && w1 <= 0x39
    then pure (fromIntegral @Word8 @Int w1 - 0x30)
    else P.fail e
  pure (w0' * 10 + w1')

-- | Universal tag number: @0x06@. Expects object identifier encoding.
oid :: Bytes -> Maybe Oid
oid = P.parseBytesMaybe objectIdentifierParser

-- | Universal tag number: @0x05@. Expects null encoding.
null :: Bytes -> Maybe ()
null b = case Bytes.length b of
  0 -> Just ()
  _ -> Nothing

-- This consumes all the way to EOF.
objectIdentifierParser :: Parser String s Oid
objectIdentifierParser = do
  P.isEndOfInput >>= \case
    True -> P.fail "oid must have length of at least 1"
    False -> pure ()
  w0 <- P.any "oid expecting first byte"
  let (v1, v2) = quotRem w0 40
      initialSize = 12
  buf0 <- P.effect (PM.newPrimArray initialSize)
  P.effect $ do
    PM.writePrimArray buf0 0 (fromIntegral @Word8 @Word32 v1)
    PM.writePrimArray buf0 1 (fromIntegral @Word8 @Word32 v2)
  let go !ix !sz !buf = P.isEndOfInput >>= \case
        True -> do
          res <- P.effect $ do
            PM.shrinkMutablePrimArray buf ix
            PM.unsafeFreezePrimArray buf
          pure (Oid res)
        False -> if ix < sz
          then do
            w <- Base128.word32 "bad oid fragment"
            P.effect (PM.writePrimArray buf ix w)
            go (ix + 1) sz buf
          else do
            let newSz = sz * 2
            newBuf <- P.effect $ do
              newBuf <- PM.newPrimArray newSz
              PM.copyMutablePrimArray newBuf 0 buf 0 sz
              pure newBuf
            go ix newSz newBuf
  go 2 initialSize buf0

-- | Universal tag number: @0x04@. Expects bitstring format.
bitString :: Bytes -> Maybe (Word8,Bytes)
bitString = P.parseBytesMaybe bitStringParser

-- | Universal tag number: @0x04@. Expects bitstring format. Expects the number of
-- padding bits to be zero.
bitStringBytes :: Bytes -> Maybe Bytes
bitStringBytes !b
  | Bytes.length b >= 1, 0 <- Bytes.unsafeHead b = Just $! Bytes.unsafeDrop 1 b
  | otherwise = Nothing

-- Always consumes all the bytes.
bitStringParser :: Parser String s (Word8,Bytes)
bitStringParser = do
  padding <- P.any "expected a padding bit count"
  if padding >= 8
    then P.fail "bitstring has more than 7 padding bits"
    else do
      bs <- P.remaining
      pure (padding,bs)

isPrintable :: Word8 -> Bool
isPrintable = \case
  0x20 -> True
  0x27 -> True
  0x28 -> True
  0x29 -> True
  0x2B -> True
  0x2C -> True
  0x2D -> True
  0x2E -> True
  0x2F -> True
  0x3A -> True
  0x3D -> True
  0x3F -> True
  w | w >= 0x41 && w <= 0x5A -> True
  w | w >= 0x61 && w <= 0x7A -> True
  w | w >= 0x30 && w <= 0x39 -> True
  _ -> False
